Allow enabling/disabling Incremental Prebuilds in Project Settings

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

Allow enabling/disabling Incremental Prebuilds in Project Settings:

- [x] Add a new Project Settings menu containing:
  - [x] The existing Configuration page
  - [x] Another page with a "Enable Incremental Prebuilds" checkbox
- [x] If it's enabled for a Project, make new Prebuilds incremental
- [x] In the Prebuild page, show when a Prebuild is incremental
- [x] Drive-by: Improve formatting of Prebuild duration
- [x] Drive-by: Correctly show 'timed out' status in Prebuild Logs

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/6727

## How to test
<!-- Provide steps to test this PR -->

1. Add a Project
2. Start a Prebuild, and wait for it to be successful
3. Only after it finished successfully, push a new commit to the same branch
4. This should trigger a new Prebuild, but the new Prebuild should not be incremental
5. Next, in the Project's Settings, enable "Use Incremental Prebuilds"
6. Push another commit to the same branch
7. The newly triggered Prebuild should be incremental

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
Allow enabling/disabling Incremental Prebuilds in Project Settings
Also improve Prebuild Logs UX (show incremental status, format duration nicely, correctly show 'timed out' status)
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

/uncc

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jankeromnes"]} -->

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by:

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["jankeromnes"]} -->

jankeromnes commented on components/dashboard/src/projects/Prebuild.tsx:
```diff
@@ -15,7 +15,9 @@ import { getGitpodService, gitpodHostUrl } from "../service/service";
 import { shortCommitMessage } from "./render-utils";
+import { Link } from "react-router-dom";
 
+// eslint-disable-next-line import/no-anonymous-default-export
```
Self-nit: No longer necessary since this rule has been disabled at the root.

jankeromnes commented on components/dashboard/src/projects/Project.tsx:
```diff
@@ -19,6 +19,7 @@ import NoAccess from "../icons/NoAccess.svg";
 import { ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
 import { openAuthorizeWindow } from "../provider-utils";
 
+// eslint-disable-next-line import/no-anonymous-default-export
```
Self-nit: Same, remove.

jankeromnes commented on components/dashboard/src/projects/ProjectSettings.tsx:
```diff
@@ -0,0 +1,76 @@
+import { getCurrentTeam, TeamsContext } from "../teams/teams-context";
+import Header from "../components/Header";
+
+// eslint-disable-next-line import/no-anonymous-default-export
```
Self-nit: Same, remove.

jankeromnes commented:
@gtsiolis Please take a look at the UX when you have time 🙏

@geropl Please take a look at the code change when you have time 🙏

geropl commented:
Starting review now...

geropl commented:
First impressions: This is set-up for confusion (Settings vs Configuration):
![image](https://user-images.githubusercontent.com/32448529/144574621-4c4abbda-1862-4c90-bcc3-6786d2e84162.png)
I had to look three times to understand where I am and to grok what the difference is.

geropl commented:
@jankeromnes  Whoops. What did I break? :grimacing:  No prebuild ran yet.
![image](https://user-images.githubusercontent.com/32448529/144575016-98b69f84-d14f-484b-ab4f-c14f6adeeec2.png)

geropl commented on components/dashboard/src/projects/Prebuilds.tsx:
```diff
@@ -257,6 +257,15 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
     }
 }
 
+function formatDuration(milliseconds: number) {
```
In the past we used [moment](https://github.com/gitpod-io/gitpod/blob/7abe78c91792c18b0c22103c890b25ea3a4cdf05/components/dashboard/package.json#L10) for this kind of work. Why not re-use that?

jankeromnes replied:
Hmm, `moment` doesn't seem that great for _formatting time differences_ specifically. (Also, their [docs](https://momentjs.com/docs/) suggest maybe not using it anymore in general.)

The primary advantage of this custom code is doing `mm:ss` when duration is less than 1 hour (e.g. `03:24`), and `h:mm:ss` when it's longer (e.g. `1:05:12`).

I guess we can achieve something similar with `moment` (using two different formats depending on >= or < 1 hour):

```ts
function formatDuration(milliseconds) {
    if (milliseconds < 1000 * 60 * 60) {
        return moment(milliseconds).format('mm:ss');
    }
    return moment(milliseconds).format('hh:mm:ss');
}
```

However (contrived edge case alert 🙈) the above code doesn't properly handle durations longer than 24 hours.

For example, for 25 hours (`milliseconds = 1000 * 60 * 60 * 25`):
- the `moment` code will return `01:00:00` (wrong!)
- the custom code will return `25:00:00` (still correct)

I guess we could handle this with yet another condition, but at this point the complexity with `moment` becomes similar to without `moment` 😋 (or, we just ignore the edge case, which is 99.99% okay but feels bad somehow).

jankeromnes replied:
Maybe like this?

```ts
function formatDuration(milliseconds) {
    const hours = Math.floor(milliseconds / (1000 * 60 * 60));
    return (hours > 0 ? `${hours}:` : '') + moment(milliseconds).format('mm:ss');
}
```

This covers all the edge cases, is short, and uses `moment`. 👍

geropl commented on components/gitpod-db/src/typeorm/entity/db-project.ts:
```diff
@@ -44,6 +44,9 @@ export class DBProject {
   @Column("simple-json", { nullable: true })
   config?: ProjectConfig;
 
+  @Column("simple-json", { nullable: true })
```
:+1:

geropl commented on components/gitpod-db/src/typeorm/migration/1638025268018-ProjectSettings.ts:
```diff
@@ -0,0 +1,21 @@
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        if (!(await columnExists(queryRunner, "d_b_project", "settings"))) {
+            await queryRunner.query("ALTER TABLE d_b_project ADD COLUMN `settings` text NULL");
```
:+1:

geropl commented on components/gitpod-db/src/typeorm/project-db-impl.ts:
```diff
@@ -70,6 +70,16 @@ export class ProjectDBImpl implements ProjectDB {
         await repo.save(project);
     }
 
+    public async setProjectSettings(projectId: string, settings: ProjectSettings): Promise<void> {
```
For these kind of methods we usually have something like `updatePartial(shape: Partial<Project> & Pick<Project, "id">)` to avoid the repetition. Having the not-exists-check in here is a good addition!

jankeromnes replied:
Sure, thanks for the hint! Went with a `DeepPartial` (so that properties like `settings` can also be Partial themselves).

geropl commented on components/server/src/projects/projects-service.ts:
```diff
@@ -233,4 +233,17 @@ export class ProjectsService {
+        if (!project) {
+            throw new Error("Project not found");
+        }
+        const settings = project.settings || {};
```
This is equvalent to
```
const settings = {
    ...(project.settings || {}),
    ...partial,
}
```
, no?
Also, we check for "project does not exist" here, again.

This is actually the 3rd time:
 - here
 - in our DB abstraction
 - typeorms `save(...)` method
Ideally we'd reduce that to one. If we rely on this in our code we should push it to the DB. And use an `update(...)` instead, which saves us the 3rd check.

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1600,6 +1601,12 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         }
     }
 
+    public async updateProjectSettings(ctx: TraceContext, projectId: string, partialSettings: Partial<ProjectSettings>): Promise<void> {
```
As mentioned above: ideally we'd have a single `udpateProjectPartial` to reduce API implementation.

geropl commented:
@jankeromnes  Finished 1st round! Nice to see this landing. :flight_arrival: 

I have no hard feelings about the UX "settings vs config". But if 1/2 hour of discussion might solve the (potential) confusion now (maybe with @gtsiolis ?) I feel that'd be well invested time.

geropl commented:
> I was sort of assuming that the Configuration page might be completely removed soon (e.g. superseded by #6921), or that we could at least move it into Settings

Ah, excellent points! :bulb:

gtsiolis commented:
Looking at this now! 👀

gtsiolis commented on components/dashboard/src/projects/ProjectSettings.tsx:
```diff
@@ -0,0 +1,76 @@
+            <div className="mt-8">
+                <h3>Incremental Prebuilds</h3>
+                <p className="text-gray-500 pb-4">When possible, use an earlier successful prebuild as a base to create new prebuilds. This can make your prebuilds significantly faster, especially if they normally take longer than 10 minutes. <a className="gp-link" href="https://www.gitpod.io/changelog/faster-incremental-prebuilds">Learn more</a></p>
+                <CheckBox title="Use Incremental Prebuilds (beta)" checked={isIncrementalPrebuildsEnabled} disabled={isLoading} onChange={toggleIncrementalPrebuilds} desc={""} />
```
**suggestion**: What do you think of using the same beta label as used in the Desktop IDE settings?

https://github.com/gitpod-io/gitpod/blob/c7c0aca070105d894643452a92bd6f79eab44a7f/components/dashboard/src/settings/Preferences.tsx#L109-L113

| Incremental Prebuilds | Desktop IDE |
|-|-|
| <img width="482" alt="Screenshot 2021-12-07 at 11 00 56 AM" src="https://user-images.githubusercontent.com/120486/144998709-fb325500-9691-4897-bc5b-18582ce7c66f.png"> | <img width="398" alt="Screenshot 2021-12-07 at 11 01 09 AM" src="https://user-images.githubusercontent.com/120486/144998721-97d2ac6b-7e67-4891-803e-72955db7a799.png"> |

gtsiolis commented on components/dashboard/src/projects/ProjectSettings.tsx:
```diff
@@ -0,0 +1,76 @@
+        <div className="app-container">
+            <div className="mt-8">
+                <h3>Incremental Prebuilds</h3>
+                <p className="text-gray-500 pb-4">When possible, use an earlier successful prebuild as a base to create new prebuilds. This can make your prebuilds significantly faster, especially if they normally take longer than 10 minutes. <a className="gp-link" href="https://www.gitpod.io/changelog/faster-incremental-prebuilds">Learn more</a></p>
```
**praise**: Thanks @nisarhassan12 for adding the individual changelog entry pages[[1](https://github.com/gitpod-io/website/pull/989)][[2](https://github.com/gitpod-io/website/issues/681)][[3](https://github.com/gitpod-io/website/issues/981)]!

gtsiolis reviewed (approved):
Thanks, @jankeromnes!

Natural next steps could be to introduce an alert component that nudges users to enable or disable incremental prebuilds based on average prebuild duration maybe of the last few prebuilds. 💭 

Approving to unblock merging but holding in case we could resolve the review comments below. Let me know if I missed anything worth discussing here or in sync.

/hold

roboquat commented:
LGTM label has been added.  <details>Git tree hash: c5ad5c121ae46a9e61b4e3e13eb219fe2bb4d75d</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-824880786" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
New changes are detected. LGTM label has been removed.

geropl commented on components/dashboard/src/Menu.tsx:
```diff
@@ -103,10 +103,14 @@ export default function Menu() {
                     title: 'Prebuilds',
                     link: `${teamOrUserSlug}/${projectName}/prebuilds`
                 },
+                {
+                    title: 'Settings',
+                    link: `${teamOrUserSlug}/${projectName}/settings`
+                },
```
> I'd expect this to always be ordered last

Agree :100:

jankeromnes commented:
Many thanks! Already addressed most comments, more to follow.

New work-in-progress Project Settings menu:

| Project Configuration | Project Settings |
| --- | --- |
| <img width="1437" alt="Screenshot 2021-12-08 at 09 50 53" src="https://user-images.githubusercontent.com/599268/145178084-7396b746-96c7-4688-9336-0dc1ecd5d815.png"> | <img width="1435" alt="Screenshot 2021-12-08 at 09 51 09" src="https://user-images.githubusercontent.com/599268/145178097-b90187e9-4af0-4ddc-aa7c-07a639c57606.png"> |

jankeromnes commented:
Many thanks for the reviews! 🙏 I've addressed all your feedback I think:

- Moved the Configuration page into the new Project Settings, and added a Project Settings side-menu
- Refactored `updateProjectSettings` to a more generic `updateProjectPartial`
- Re-styled the Checkbox
- Fixed a few more bugs

@geropl Could you please take another look at the code when you have time? 🙏

@gtsiolis Could you please check the UX again, in particular the new Project Settings menu? 🙏

gtsiolis commented:
Looking at this now! 👀

gtsiolis commented on components/dashboard/src/projects/ProjectSettings.tsx:
```diff
@@ -0,0 +1,94 @@
+
+    return <PageWithSubMenu subMenu={getProjectSettingsMenu(project, team)} title="Settings" subtitle={`Manage settings for your project ${project?.name}`}>
+        <h3>Incremental Prebuilds</h3>
+        <CheckBox
```
**issue**: Could we use `max-w-2xl` for the checbox text here to limit the width of this element as we do with similar sections like Account settings?

**fyi**: The grid layout from earlier mockups is not accurately translated in the dashboard yet but will become more relevant once we pick up issues related to responsive design of the dashboard. Posting below an early draft of the responsive design of the grid. Using `max-w-2xl` here is a workaround to approach as much as possible the designs we've been using so far to create visual balance when needed. Feedback is welcome!

| Grid layout in large, medium, small and extra small viewports |
|-|
| <img width="1705" alt="Screenshot 2021-12-09 at 10 27 52 PM" src="https://user-images.githubusercontent.com/120486/145470842-f756fb77-a684-4a61-bf20-1fe06b1a4f49.png"> |

This is also an issue with other parts of the product, see [relevant discussion](https://gitpod.slack.com/archives/C02CBKZEVS8/p1638795733242600?thread_ts=1638366112.229500&cid=C02CBKZEVS8) (internal). Cc @gitpod-io/engineering-ide

gtsiolis commented on components/dashboard/src/teams/TeamSettings.tsx:
```diff
@@ -44,15 +54,8 @@ export default function TeamSettings() {
-
     return <>
-        <PageWithSubMenu subMenu={settingsMenu} title='General' subtitle='Manage general team settings.'>
+        <PageWithSubMenu subMenu={getTeamSettingsMenu(team)} title="Settings" subtitle="Manage general team settings.">
```
**suggestion**: What do you think of :a: renaming this menu _General_ to follow the pattern we use for team settings and :b: reordering this so it's always the first sub menu entry?

jankeromnes replied:
🅱️: Sure 👍 happy to make this the top-level entry in the left submenu.

🅰️: I actually find it confusing when the BIG TITLE at the top of the page is different from the Tab I just clicked on: When I click on "Settings", I expect to see "Settings" (and not "General" -- I don't know what that is yet, so I think I mis-clicked).

So I had actually renamed the other "General" title in this PR 🙈

| Project Settings | Team Settings |
| --- | --- |
| <img width="1437" alt="Screenshot 2021-12-13 at 15 58 45" src="https://user-images.githubusercontent.com/599268/145835979-75166548-28b5-496f-a5b4-41a1cde60f8a.png"> | <img width="1436" alt="Screenshot 2021-12-13 at 15 58 10" src="https://user-images.githubusercontent.com/599268/145836096-c0f07bdc-2cc7-48c1-b053-e829a07c9076.png"> |

You'll notice that I've also renamed the left submenu entry to `Project` and `Team` respectively, because I felt that made sense. However, I'm happy to rename both of these to `General` (but keep `Settings` as the BIG TITLE at the top).

What do you think? 🏓

gtsiolis replied:
:b: DEAL 🤝 

:a: Valid concern! This pattern was inherited for team settings from the personal account settings where is makes more sense to differentiate multiple options, see [`/settings`](https://gitpod.io/settings). Keeping the _Settings_ title could make sense for now that there're only two options but could be an issue when project and team settings grow over time. For example, see https://github.com/gitpod-io/gitpod/issues/4456, https://github.com/gitpod-io/gitpod/issues/5067, https://github.com/gitpod-io/gitpod/issues/7101, https://github.com/gitpod-io/gitpod/issues/6706#issuecomment-970649837, etc.

gtsiolis commented on components/dashboard/src/teams/TeamSettings.tsx:
```diff
@@ -44,15 +54,8 @@ export default function TeamSettings() {
-        <PageWithSubMenu subMenu={settingsMenu} title='General' subtitle='Manage general team settings.'>
+        <PageWithSubMenu subMenu={getTeamSettingsMenu(team)} title="Settings" subtitle="Manage general team settings.">
             <h3>Delete Team</h3>
             <p className="text-base text-gray-500 pb-4">Deleting this team will also remove all associated data with this team, including projects and workspaces. Deleted teams cannot be restored!</p>
```
**issue**: This one also needs to be limited in width with `max-w-2xl`

gtsiolis commented on components/dashboard/src/projects/ConfigureProject.tsx:
```diff
@@ -244,9 +227,8 @@ export default function () {
+        <div className="flex space-x-4">
             <div className="flex-1 h-96 rounded-xl overflow-hidden relative flex flex-col">
                 <div className="flex bg-gray-50 dark:bg-gray-800 border-b border-gray-200 dark:border-gray-600 px-6 pt-3">
                     <TabMenuItem name=".gitpod.yml" selected={selectedEditor === '.gitpod.yml'} onClick={() => setSelectedEditor('.gitpod.yml')} />
```
**question**: I could be missing something but do we think it makes sense to remove the latest prebuild block here and only keep the configuration n now that we've removed this from the onboarding flow? Later on we could also handle cases where no configuration is added yet.

**thought**: Eventually we will need to separate project addition and project configuration steps to that these can be done separastely.

jankeromnes replied:
Sure, I'm happy to remove the "Prebuild Runner", i.e. the right part of the split screen, and keep only the editor.

However, this feels like a follow-up issue (because I think we also need to fix the copy).

gtsiolis commented on components/dashboard/src/projects/ProjectSettings.tsx:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    return <PageWithSubMenu subMenu={getProjectSettingsMenu(project, team)} title="Settings" subtitle={`Manage settings for your project ${project?.name}`}>
```
**suggestion**: We already surface the project name in the top navigation menu. Shallow we remove it from here?

```suggestion
    return <PageWithSubMenu subMenu={getProjectSettingsMenu(project, team)} title="Settings" subtitle="Manage project settings and configuration">
```

jankeromnes replied:
Good point, we already see the slug at the top (which should actually be the name, now that I think of it).

Will fix both. 😁

gtsiolis reviewed (approved):
Great changes, @jankeromnes! 🧙 

Excited to see this feature becoming part of project settings! 🙌 

Approving to unblock merging but holding because of some comments below and before @geropl take a closer look at the code changes. 🏀  

/hold

roboquat commented:
LGTM label has been added.  <details>Git tree hash: b0beeec6cfe09f3bfe3f800bf346c69055dd3f28</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

gtsiolis commented:
**praise**: @jankeromnes I wanted to drop by again and say thanks again for always using the PR description for breaking down the changes of each PR and including clear steps to test a PR. 🌟

geropl commented on components/gitpod-db/src/typeorm/project-db-impl.ts:
```diff
@@ -60,14 +60,13 @@ export class ProjectDBImpl implements ProjectDB {
         }
-        project.config = config;
-        await repo.save(project);
+        await repo.update(partialProject.id, partialProject);
```
:+1:

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1713,6 +1702,35 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+        await this.guardProjectOperation(user, partialProject.id, "update");
+
+        const partial: PartialProject = { id: partialProject.id };
+        const allowedFields: (keyof Project)[] = ['settings']; // Don't add 'config' here! Please use `setProjectConfiguration` instead to parse & validate configs
```
:100: Thx! IMO we should far often be more careful like this, especially when adding super-powerful methods that might change over time. :slightly_smiling_face:

geropl commented on components/gitpod-protocol/src/teams-projects-protocol.ts:
```diff
@@ -55,6 +61,8 @@ export namespace Project {
     }
 }
 
+export type PartialProject = DeepPartial<Project> & Pick<Project, 'id'>;
```
:+1:

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1713,6 +1702,35 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+    }
+
+    public async updateProjectPartial(ctx: TraceContext, partialProject: PartialProject): Promise<void> {
+        const user = this.checkUser("updateProjectPartial");
```
nit: let's add a `traceAPIParams` like so
```suggestion
        traceAPIParams(ctx, { partialProject: { id: partialProject.id, settings: partialProject.settings });   // censor everything irrelevant

        const user = this.checkUser("updateProjectPartial");
```

geropl commented:
/lgtm

@jankeromnes I keep the `hold` in case you want to tackle https://github.com/gitpod-io/gitpod/pull/7031#discussion_r766471777 within this PR.

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#issuecomment-990735262" title="LGTM">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

roboquat commented:
New changes are detected. LGTM label has been removed.

jankeromnes commented:
/unhold

geropl commented:
/lgtm

Code is fine, and tested last time! :+1:

roboquat commented:
LGTM label has been added.  <details>Git tree hash: ec9da6fb245e4f3595a832f5ac2fc39bfce1d6ec</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7031#issuecomment-993243906" title="LGTM">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7031#pullrequestreview-828154934" title="Approved">gtsiolis</a>*

Associated issue: *#6727*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,gtsiolis]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,gtsiolis]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
